src/services/RPCSelector.ts

```diff
@@ -0,0 +1,69 @@
+import { JsonRpcProvider, Block } from '@ethersproject/providers';
+import { isTimestampMoreRecentThan } from '../utils/helpers';
+
+interface ProviderComparand {
+  blockNumber: number;
+  url: string;
+}
+
+class RPCSelector {
+  readonly urls: string[];
+  readonly preferredProviderUrl: string;
+  readonly rpcRequestTimeout: number;
+  readonly maxBlockAge: number;
+
+  constructor(urls: string | string[], rpcRequestTimeout = 15000, maxBlockAge = 60) {
+    this.urls = typeof urls === 'string' ? urls.split(',') : urls;
+    this.preferredProviderUrl = this.urls[0];
+    this.rpcRequestTimeout = rpcRequestTimeout;
+    this.maxBlockAge = maxBlockAge;
+  }
+
+  async apply(): Promise<string> {
+    if (this.urls.length === 1 || (await this.isPreferredProviderUpToDate())) {
+      return this.preferredProviderUrl;
+    } else {
+      const providers = await Promise.all(this.getProviders(this.urls.slice(1)));
+      return this.getMostUpToDateProvider(providers);
+    }
+  }
+
+  private async isPreferredProviderUpToDate(): Promise<boolean> {
+    try {
+      const provider = new JsonRpcProvider(this.preferredProviderUrl);
```

Comment:
what if we set web socket url?

Reply:
I asked you this on discord 9 days ago. We can add support to WebSockets but it will take a little more time, as usual.